ff2png.1
.Dd 2024-03-28
.Dt FF2PNG 1
.Os farbfeld-utils
.Sh NAME
.Nm ff2png
.Nd convert farbfeld to PNG
.Sh SYNOPSIS
.Nm
<?options...?>
.Sh DESCRIPTION
.Nm
reads a
.Xr farbfeld 5
image from stdin, converts it to PNG (16-bit RGBA) and writes the result to stdout.
.Pp
In case of an error
.Nm
writes a diagnostic message to stderr.
.Sh EXIT STATUS
.Bl -tag -width Ds
.It 0
Image processed successfully.
.It 1
An error occurred.
.El
.Sh OPTIONS
The encoder can optionally take command-line arguments.

If none are specified it will be using express mode, which uses the default settings for everything.
If at least one is specified it uses custom mode with customizable settings.

Each argument is a separate setting. Each setting consists of one letter (or other character)
followed by the value (with no delimiter).

Valid settings are:

.Bl -tag -width Ds
.It =
Use a key=value pair to enter that text as metadata into the encoded PNG file. Use this setting
multiple times for multiple text metadata.
.It +
No value is used. All remaining arguments specify the palette. This palette does not support
16-bits per channel. Mandatory if c3 is specified; otherwise it is only a suggested palette.
.It @
Specify a filename, and will read that file and copy all of the safe chunks.
.It a
Add a metadata chunk specifying the version of LodePNG used if this setting is set to 1.
.It b
Specify zero to disable DEFLATE compression. You can also specify one to use a fixed Huffman tree.
The default setting is two.
.It c
Override the auto detection of colour mode. Values can be: 0=greyscale, 2=RGB, 3=indexed,
4=greyscale with alpha, 6=RGB with alpha. You can then optionally include a comma and a second
number which is the bit depth to use.
.It e
Some extra settings (using an unofficial version of LodePNG). These are up to five numbers with
commas in between. The first number is the DEFLATE block size (specify -1 to use a single block),
the second number is the max chain length, the third number is the max lazy match, the fourth
number is a bit field described below, and the fifth number is the maximum distance for matches
of length 3 (default 4096). For the first three settings, 0 means use default. For the block size,
if c is already specified before this one, can optionally consist of "s" followed by a number of
scanlines.
.It f
Specify the filter strategy. Values can be: 0=no filter, 1=minimum sum, 2=Shannon entropy, 3=brute
force, 10=two line Shannon entropy. The default setting is to use 1 for pictures with more than
256 colours and 0 otherwise.
.It g
Specify gamma setting of PNG file. The actual contents of the picture is not altered; it only adds
the extra data into the file which specifies what gamma setting this picture is corresponded to.
The value of this setting is allowed to include a fractional part, unlike the others which are
only integers.
.It i
Set interlacing. Specify zero to disable interlacing (default) or one to enable Adam7 interlacing.
.It k
If you are copying the chunks from other file, then you can specify what stuff can be kept:
0=safe only, 1=a few more, 9=everything.
.It l
Specify zero to disable lazy matching, or one to enable lazy matching (the default setting). Lazy
matching result better compression but is slower.
.It m
Minimum LZ77 length, normally 3. Some pictures will compress better if it is 6 instead.
.It n
Specify the maximum LZ77 length to search; stop after searching this length. The default is 128,
and the maximum is 258.
.It p
Specify physical size as three numbers with commas in between. First is horizontal resolution,
second is vertical resolution, and the third specifies the units of measurement.
.It q
Manual filter setting for all scanlines. The setting can be: 0=none, 1=subtract left, 2=subtract
above, 3=average left and above, 4=Paeth.
.It t
Try to optimize transparent pixels. The setting can be 0 to not do so, or 1 to change the filtered
value to 0 for transparent pixels; other numbers use other strategies. If nonzero, the colour of
transparent pixels is not necessarily the same as the input. (Note: Experimental feature)
.It u
Specify zero to disable LZ77 compression.
.It w
The window size for LZ77 compression, which is a power of two from 1 to 32768. The default value
is 2048, although using 32768 can usually result in improved compression.
.It z
Specify one to compress text metadata, or zero to disable.
.El

Note that it will automatically set the bit depth and/or palette as needed in order to encode the
picture losslessly. Which settings result in best compression can depend on the picture, and are
not necessarily the same as the recommendations specified above.

The bit field described for the fourth number for "e" is as follows:

.Bl -tag -width Ds
.It 1
Consider strings of zeros (default).
.It 2
Make the final block larger so that a small block isn't encoded.
.El

Also see RFC 2083.
.Sh REFERENCES
https://lodev.org/lodepng/
.Sh EXAMPLES
If you want to reduce bit depth to 8-bits-per-channel and force fully
transparent pixels to become zero then the last example command shows how to
do this by combining
.Xr ff-poster 1
and
.Xr ff-transopt 1
.
.Pp
$
.Nm
< image.ff > image.png
.Pp
$ bunzip2 < image.ff.bz2 |
.Nm
> image.png
.Pp
$ ff-poster 8 < image.ff | ff-transopt |
.Nm
> image.png
.Sh SEE ALSO
.Xr png2ff 1 ,
.Xr ff-poster 1 ,
.Xr ff-transopt 1 ,
.Xr bzip2 1 ,
.Xr farbfeld 5


ff2png.c
#if 0
gcc -s -O2 -o ~/bin/ff2png ff2png.c lodepng.c
exit
#endif
/*
	Convert Farbfeld to PNG
	(public domain)
*/

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include "lodepng.h"

static FILE*in_png;
static int keep_which;

static void
usage(void)
{
	fprintf(stderr,
		"Usage: ff2png <?options...?>                                                  \n"
		"                                                                              \n"
		"Convert farbfeld to PNG. ff2png reads a farbfeld image from stdin, converts it\n"
		"to PNG (16-bit RGBA) and writes the result to stdout.                         \n"
		"                                                                              \n"
		"This encoder can optionally use command-line arguments. If none are specified \n"
		"it will be using express mode which uses the default settings for everything. \n"
		"                                                                              \n"
		"Refer to the man page for command line options.                               \n"
		"                                                                              \n"
		"If you want to reduce bit depth to 8-bits-per-channel and force fully         \n"
		"transparent pixels to become zero then the last example command shows how to  \n"
		"do this by combining ff-poster and ff-transopt.                               \n"
		"                                                                              \n"
		"Example usage:                                                                \n"
		"   $ ff2png < image.ff > image.png                                            \n"
		"   $ bunzip2 < image.ff.bz2 | ff2png > image.png                              \n"
		"   $ ff-poster 8 < image.ff | ff-transopt | ff2png > image.png                \n"
		"\n"
	);
	exit(1);
}

static void
do_in_png(const unsigned char *out)
{
	unsigned char buf[8];
	const unsigned char*ptr=out+8;
	unsigned long len;
	if (fread(buf,1,8,in_png)!=8 || memcmp(out,buf,8)) {
		fprintf(stderr,"Not a PNG file\n");
		exit(1);
	}
	fwrite(out,1,8,stdout);
loop1:
	len=((ptr[0]<<24)|(ptr[1]<<16)|(ptr[2]<<8)|ptr[3])+12;
	fwrite(ptr,1,len,stdout);
	if (!memcmp(ptr+4,"IEND",4)) return;
	ptr+=len;
	if (memcmp(ptr-len+4,"IHDR",4)) goto loop1;
loop2:
	if (fread(buf,1,8,in_png)!=8) {
		perror(0);
		exit(1);
	}
	len=((buf[0]<<24)|(buf[1]<<16)|(buf[2]<<8)|buf[3])+4;
	if ((buf[7]&32) || (keep_which>1 && (buf[4]&32)) || (keep_which==1 && !(memcmp(buf+4,"cHRM",4) && memcmp(buf+4,"gAMA",4)))) {
		fwrite(buf,1,8,stdout);
		while (len--) putchar(fgetc(in_png));
	} else {
		fseek(in_png,len,SEEK_CUR);
	}
	if (memcmp(buf+4,"IEND",4)) goto loop2;
	goto loop1;
}

static void
make_gamma(LodePNGState*st, int ga)
{
	unsigned char buf[4];
	unsigned char*chunk=0;
	size_t len=0;
	buf[0]=ga>>24;
	buf[1]=ga>>16;
	buf[2]=ga>>8;
	buf[3]=ga>>0;
	lodepng_chunk_create(&chunk,&len,4,"gAMA",buf);
	lodepng_chunk_append(st->info_png.unknown_chunks_data,st->info_png.unknown_chunks_size,chunk);
}

int
main(int argc, char **argv)
{
	unsigned char head[16];
	unsigned char*in;
	unsigned char*out;
	unsigned w,h,e;
	if (argc > 1 && (!strcmp(argv[1],"-h") || !strcmp(argv[1],"--help"))) {
		usage();
	}
	size_t rsize=fread(head,1,16,stdin);

	if (rsize!=16 || memcmp("farbfeld",head,8)) {
		fprintf(stderr,"Not farbfeld\n");
		return 1;
	}
	w=(head[8]<<24)|(head[9]<<16)|(head[10]<<8)|head[11];
	h=(head[12]<<24)|(head[13]<<16)|(head[14]<<8)|head[15];
	in=malloc(w*h*8);
	if (!in) {
		fprintf(stderr,"Allocation failed\n");
		return 1;
	}
	rsize=fread(in,8,w*h,stdin);
	if (argc<2) {
		if (e=lodepng_encode_memory(&out,&rsize,in,w,h,LCT_RGBA,16)) {
			fprintf(stderr,"LodePNG: %s\n",lodepng_error_text(e));
			return 1;
		}
	} else {
		LodePNGState st;
		char t[80];
		int i;
		unsigned char*buf;
		lodepng_state_init(&st);
		st.info_raw.colortype=LCT_RGBA;
		st.info_raw.bitdepth=16;
		for (e=1; e<argc; e++) {
			switch(argv[e][0]) {
				case '=':
					strncpy(t,argv[e]+1,79);
					t[strcspn(t,"=")]=0;
					lodepng_add_text(&st.info_png,t,argv[e]+strlen(t)+2);
					break;
				case '+':
					st.info_png.color.palette=buf=malloc(1024);
					if (!buf) {
						fprintf(stderr,"Allocation failed\n");
						return 1;
					}
					st.encoder.force_palette=1;
					e++;
					goto palet;
				case '@':
					in_png=fopen(argv[e]+1,"r");
					if (!in_png) {
						perror(argv[e]+1);
						return 1;
					}
					break;
				case 'a':
					st.encoder.add_id=argv[e][1]-'0';
					break;
				case 'b':
					st.encoder.zlibsettings.btype=argv[e][1]-'0';
					break;
				case 'c':
					st.encoder.auto_convert=0;
					st.info_png.color.colortype=argv[e][1]-'0';
					st.info_png.color.bitdepth=argv[e][2]==','?strtol(argv[e]+3,0,0):8;
					st.info_png.color.palettesize=0;
					st.info_png.color.key_defined=0;
					break;
				case 'e':
					if(argv[e][1]=='s' && !st.encoder.auto_convert) {
						i="\x01\x00\x03\x01\x02\x00\x04"[st.info_png.color.colortype];
						i=(i*w*8LL)/st.info_png.color.bitdepth+1;
						st.encoder.zlibsettings.blocksize=i*strtol(argv[e]+2,(char**)&buf,0);
					} else {
						st.encoder.zlibsettings.blocksize=strtol(argv[e]+1,(char**)&buf,0);
					}
					if (*buf==',') st.encoder.zlibsettings.maxchainlength=strtol(buf+1,(char**)&buf,0);
					if (*buf==',') st.encoder.zlibsettings.maxlazymatch=strtol(buf+1,(char**)&buf,0);
					if (*buf==',') st.encoder.zlibsettings.usezeros=strtol(buf+1,(char**)&buf,0);
					if (*buf==',') st.encoder.zlibsettings.too_far=strtol(buf+1,0,0);
					break;
				case 'f':
					st.encoder.filter_palette_zero=0;
					st.encoder.filter_strategy=strtol(argv[e]+1,0,0);
					break;
				case 'g':
					make_gamma(&st,(int)(strtod(argv[e]+1,0)*100000.0+0.45));
					break;
				case 'i':
					st.info_png.interlace_method=argv[e][1]-'0';
					break;
				case 'k':
					keep_which=argv[e][1]-'0';
					break;
				case 'l':
					st.encoder.zlibsettings.lazymatching=argv[e][1]-'0';
					break;
				case 'm':
					st.encoder.zlibsettings.minmatch=strtol(argv[e]+1,0,0);
					break;
				case 'n':
					st.encoder.zlibsettings.nicematch=strtol(argv[e]+1,0,0);
					break;
				case 'p':
					st.info_png.phys_defined=1;
					st.info_png.phys_x=strtol(argv[e]+1,(char**)&buf,0);
					if (*buf==',') st.info_png.phys_y=strtol(buf+1,(char**)&buf,0);
					if (*buf==',') st.info_png.phys_unit=strtol(buf+1,0,0);
					break;
				case 'q':
					st.encoder.filter_palette_zero=0;
					st.encoder.filter_strategy=LFS_PREDEFINED;
					if (!(buf=malloc(h))) {
						fprintf(stderr,"Allocation failed\n");
						return 1;
					}
					if (argv[e][1]=='-') {
						for (i=0; i<h; i++) buf[i]=i?0:2;
					} else {
						for (i=0; i<h; i++) buf[i]=argv[e][1]-'0';
					}
					st.encoder.predefined_filters=buf;
					break;
				case 't':
					st.encoder.trans_filter=strtol(argv[e]+1,0,0);
					break;
				case 'u':
					st.encoder.zlibsettings.use_lz77=argv[e][1]-'0';
					break;
				case 'w':
					st.encoder.zlibsettings.windowsize=strtol(argv[e]+1,0,0);
					break;
				case 'z':
					st.encoder.text_compression=argv[e][1]-'0';
					break;
				default:
					if (argv[e][0]) {
						fprintf(stderr,"Unknown option: %c\n",argv[e][0]);
						return 1;
					}
			}
		}
		palet:
		for (i=0; e<argc; e++) {
			if (i==256) {
				fprintf(stderr,"Too many palette entries\n");
				return 1;
			}
			if (strlen(argv[e])==8) {
				sscanf(argv[e],"%2hhX%2hhX%2hhX%2hhX",buf+(i<<2)+0,buf+(i<<2)+1,buf+(i<<2)+2,buf+(i<<2)+3);
			} else if (strlen(argv[e])!=6) {
				fprintf(stderr,"Invalid palette specification\n");
				return 1;
			} else {
				sscanf(argv[e],"%2hhX%2hhX%2hhX",buf+(i<<2)+0,buf+(i<<2)+1,buf+(i<<2)+2);
				buf[(i<<2)|3]=255;
			}
			st.info_png.color.palettesize=++i;
		}
		if (st.encoder.zlibsettings.btype!=2 || st.info_png.interlace_method) st.encoder.zlibsettings.custom_deflate=0;
		if (e=lodepng_encode(&out,&rsize,in,w,h,&st)) {
			fprintf(stderr,"LodePNG: %s\n",lodepng_error_text(e));
			return 1;
		}
	}
	if (in_png) do_in_png(out); else fwrite(out,1,rsize,stdout);
	return 0;
}
